src/encode.rs

```diff
@@ -72,31 +71,148 @@ pub fn encode_buffered(input: Slipmux) -> Vec<u8> {
+/// UART output buffer, which might be reused as soon as some data is flushed out.
+pub struct ChunkedEncoder<'input> {
+    header: &'static [u8],
+    data: &'input [u8],
+    // Note that this is often initialized to 2 to not serialize any FCS
+    fcs_cursor: u8,
+    fcs: [u8; 2],
+    slip: Option<Encoder>,
+}
+
+impl<'input> ChunkedEncoder<'input> {
+    /// Creates a new chunked encoder.
+    ///
+    /// This will encode the data based on the `FrameType` over any number of calls to
+    /// [`Self::encode_chunk()`] into a buffer provided there.
+    #[must_use]
+    pub fn new(ftype: FrameType, data: &'input [u8]) -> Self {
+        let header: &[u8] = match ftype {
+            FrameType::Diagnostic => &[Constants::DIAGNOSTIC],
+            FrameType::Configuration => &[Constants::CONFIGURATION],
+            FrameType::Ip => &[],
+        };
+
+        let (fcs, fcs_cursor) = if matches!(ftype, FrameType::Configuration) {
             let fcs = fcs16_part(CONF_FCS16, data);
-            let fcs = fcs16_finish(fcs);
-            totals += slip.encode(&[Constants::CONFIGURATION], buffer).unwrap();
-            totals += slip.encode(data, &mut buffer[totals.written..]).unwrap();
-            totals += slip
-                .encode(&fcs.to_le_bytes(), &mut buffer[totals.written..])
-                .unwrap();
+            (fcs16_finish(fcs).to_le_bytes(), 0)
+        } else {
+            ([0, 0], 2)
+        };
+
+        Self {
+            header,
+            data,
+            fcs_cursor,
+            fcs,
+            slip: Some(Encoder::new()),
+        }
+    }
+
+    /// Non-empty next piece of data that should be encoded, or None if all data has been encoded
+    /// (but probably not the End marker).
+    fn slice_to_encode(&self) -> Option<&[u8]> {
+        if !self.header.is_empty() {
+            Some(self.header)
+        } else if !self.data.is_empty() {
+            Some(self.data)
+        } else if (self.fcs_cursor as usize) < self.fcs.len() {
+            Some(&self.fcs[self.fcs_cursor as usize..])
+        } else {
+            None
+        }
+    }
+
+    /// Advance whichever slice was just selected in [`Self::slice_to_encode()`] by some amount of
+    /// bytes.
+    fn advance_slice(&mut self, amount: usize) {
+        if !self.header.is_empty() {
+            self.header = &self.header[amount..];
+        } else if !self.data.is_empty() {
+            self.data = &self.data[amount..];
+        } else {
+            self.fcs_cursor += amount as u8;
         }
-        FrameType::Ip => {
-            totals += slip.encode(data, &mut buffer[totals.written..]).unwrap();
+    }
+
+    /// Writes some of the data into the output buffer.
+    ///
+    /// Returns the number of bytes written.
+    ///
+    /// Unlike many other write-style methods, this does *not* fill up the buffer to the last byte
+    /// except for the last block; this simplifies encoding escaped bytes.
```

Comment:
But it does fill the buffer up the buffer to the last byte? Not if the buffer is very big but if the buffer is smaller than the data, it will always fill the buffer to the limit.